File: trace-agent-bootstrap/src/test/java/com/vmlens/trace/agent/bootstrap/interleave/run/inttest/VolatileFieldAccessTestBuilder.java
package com.vmlens.trace.agent.bootstrap.interleave.run.inttest;

import com.vmlens.trace.agent.bootstrap.MemoryAccessType;
import com.vmlens.trace.agent.bootstrap.interleave.interleaveActionImpl.VolatileAccess;
import com.vmlens.trace.agent.bootstrap.interleave.interleaveActionImpl.volatileaccesskey.VolatileFieldAccessKey;
import com.vmlens.trace.agent.bootstrap.interleave.run.InterleaveAction;

public class VolatileFieldAccessTestBuilder {

    private final int fieldId;

    public VolatileFieldAccessTestBuilder(int fieldId) {
        this.fieldId = fieldId;
    }

    public static VolatileFieldAccessTestBuilder firstVolatileField() {
        return new VolatileFieldAccessTestBuilder(1);
    }

    public InterleaveActionTestFactory read() {
        return new VolatileFieldAccessFactory(fieldId, MemoryAccessType.IS_READ);
    }

    public InterleaveActionTestFactory write() {
        return new VolatileFieldAccessFactory(fieldId, MemoryAccessType.IS_WRITE);
    }

    private static class VolatileFieldAccessFactory implements InterleaveActionTestFactory {

        private final int fieldId;
        private final int operation;

        public VolatileFieldAccessFactory(int fieldId, int operation) {
            this.fieldId = fieldId;
            this.operation = operation;
        }

        @Override
        public InterleaveAction create(int threadIndex) {
            return new VolatileAccess(threadIndex, new VolatileFieldAccessKey(fieldId,6L), operation);
        }
    }

}

File: sync-bug/src/test/scala/com/anarsoft/race/detection/createpartialordersyncaction/inttest/builder/VolatileFieldAccessTestBuilder.scala
package com.anarsoft.race.detection.createpartialordersyncaction.inttest.builder

import com.anarsoft.race.detection.createpartialordersyncaction.SyncActionEventWithCompareType
import com.anarsoft.race.detection.createpartialordersyncaction.inttest.generic.ReadOrWriteEventBuilder
import com.anarsoft.race.detection.event.gen.VolatileFieldAccessEventGen
import com.anarsoft.race.detection.event.interleave.VolatileFieldAccessEvent
import com.vmlens.trace.agent.bootstrap.MemoryAccessType

class VolatileFieldAccessTestBuilder(val fieldId  : Int, val objectHashCode  : Long) 
     extends AbstractTestBuilder with ReadOrWriteEventBuilder[VolatileFieldAccessEvent]{

  def read(threadIndex : Int): VolatileFieldAccessEvent =  event(threadIndex,MemoryAccessType.IS_READ)

  def write(threadIndex : Int): VolatileFieldAccessEvent = event(threadIndex,MemoryAccessType.IS_WRITE)

  private def event( threadIndex : Int, operation : Int) : VolatileFieldAccessEvent = {
    val pos = nextPosition(threadIndex);
    new VolatileFieldAccessEventGen(threadIndex,
      bytecodePosition,
      fieldId,
      methodCounter,
      methodId,
      operation,
      objectHashCode,
      loopId,
       runId,
      pos.runPosition
    )
  }

}
